File: golang/lib/execution/kurtosis_lambda_executor.go
/*
 * Copyright (c) 2020 - present Kurtosis Technologies LLC.
 * All Rights Reserved.
 */

package execution

import (
	"github.com/kurtosis-tech/kurtosis-client/golang/kurtosis_core_rpc_api_bindings"
	"github.com/kurtosis-tech/kurtosis-client/golang/lib/networks"
	"github.com/kurtosis-tech/kurtosis-client/golang/lib/services"
	"github.com/kurtosis-tech/kurtosis-lambda-api-lib/golang/kurtosis_lambda_docker_api"
	"github.com/kurtosis-tech/kurtosis-lambda-api-lib/golang/kurtosis_lambda_rpc_api_bindings"
	"github.com/kurtosis-tech/kurtosis-lambda-api-lib/golang/kurtosis_lambda_rpc_api_consts"
	"github.com/kurtosis-tech/minimal-grpc-server/server"
	"github.com/palantir/stacktrace"
	"google.golang.org/grpc"
	"os"
	"time"
)

const (
	grpcServerStopGracePeriod = 5 * time.Second
)

type KurtosisLambdaExecutor struct {
	configurator KurtosisLambdaConfigurator
}

func NewKurtosisLambdaExecutor(configurator KurtosisLambdaConfigurator) *KurtosisLambdaExecutor {
	return &KurtosisLambdaExecutor{configurator: configurator}
}

func (executor KurtosisLambdaExecutor) Run() error {

	serializedCustomParams, err := getEnvVar(kurtosis_lambda_docker_api.SerializedCustomParamsEnvVar, "the serialized custom params that the Lambda will consume")
	if err != nil {
		return stacktrace.Propagate(err, "An error occurred when trying to get the serialized custom params environment variable")
	}

	lambda, err := executor.configurator.ParseParamsAndCreateKurtosisLambda(serializedCustomParams)
	if err != nil {
		return stacktrace.Propagate(err, "An error occurred parsing the serialized custom params and creating the Lambda")
	}

	apiContainerSocket, err := getEnvVar(kurtosis_lambda_docker_api.ApiContainerSocketEnvVar, "the socket value used in API container connection")
    if err != nil {
		return stacktrace.Propagate(err, "An error occurred when trying to get the API container socket environment variable")
	}

	// TODO SECURITY: Use HTTPS to verify we're hitting the correct API container
	conn, err := grpc.Dial(apiContainerSocket, grpc.WithInsecure())
	if err != nil {
		return stacktrace.Propagate(err, "An error occurred dialling the API container at '%v'", apiContainerSocket)
	}

	apiClient := kurtosis_core_rpc_api_bindings.NewApiContainerServiceClient(conn)
	networkCtx := networks.NewNetworkContext(
		apiClient,
		map[services.FilesArtifactID]string{},
		kurtosis_lambda_docker_api.ExecutionVolumeMountpoint,
	)

	lambdaServiceServer := NewKurtosisLambdaServiceServer(lambda, networkCtx)
	lambdaServiceRegistrationFunc := func(grpcServer *grpc.Server) {
		kurtosis_lambda_rpc_api_bindings.RegisterLambdaServiceServer(grpcServer, lambdaServiceServer)
	}

	lambdaServer := server.NewMinimalGRPCServer(
		kurtosis_lambda_rpc_api_consts.ListenPort,
		kurtosis_lambda_rpc_api_consts.ListenProtocol,
		grpcServerStopGracePeriod,
		[]func(desc *grpc.Server){
			lambdaServiceRegistrationFunc,
		},
	)
	if err := lambdaServer.Run(); err != nil {
		return stacktrace.Propagate(err, "An error occurred running the Lambda server")
	}

	return nil
}

// ====================================================================================================
//                                       Private helper functions
// ====================================================================================================
func getEnvVar(envVarName string, envVarDescription string) (string, error) {
	envVarValue, found := os.LookupEnv(envVarName)

	if !found {
		return "", stacktrace.NewError("Expected an '%v' environment variable containing '%v', but none was found", envVarName, envVarDescription)
	}
	if envVarValue == "" {
		return "", stacktrace.NewError("The '%v' serialized custom params environment variable was defined, but is emptystring", envVarName)
	}

	return envVarValue, nil
}

File: typescript/src/lib/execution/kurtosis_lambda_executor.ts
import { ApiContainerServiceClient, NetworkContext } from "kurtosis-core-api-lib";
import { err, ok, Result } from "neverthrow";
import { API_CONTAINER_SOCKET_ENV_VAR, EXECUTION_VOLUME_MOUNTPOINT, SERIALIZED_CUSTOM_PARAMS_ENV_VAR } from "../../kurtosis_lambda_docker_api/kurtosis_lambda_docker_api";
import { KurtosisLambda } from "../kurtosis_lambda/kurtosis_lambda";
import { KurtosisLambdaConfigurator } from "./kurtosis_lambda_configurator";
import * as grpc from 'grpc';
import { ILambdaServiceServer, LambdaServiceService } from "../../kurtosis_lambda_rpc_api_bindings/api_lambda_service_grpc_pb";
import { MinimalGRPCServer, TypedServerOverride } from "minimal-grpc-server";
import { LISTEN_PORT } from "../../kurtosis_lambda_rpc_api_consts/kurtosis_lambda_rpc_api_consts";

const GRPC_SERVER_STOP_GRACE_PERIOD_SECONDS: number = 5;

export class KurtosisLambdaExecutor {
    private readonly configurator: KurtosisLambdaConfigurator;

    constructor(configurator: KurtosisLambdaConfigurator) {
        this.configurator = configurator;
    }

    async run(): Promise<Result<null, Error>> {
        const getSerializedCustomParamsResult: Result<string, Error> = KurtosisLambdaExecutor.getEnvVar(SERIALIZED_CUSTOM_PARAMS_ENV_VAR, "the serialized custom params that the Lambda will consume");
        if (getSerializedCustomParamsResult.isErr()) {
            return err(getSerializedCustomParamsResult.error);
        }
        const serializedCustomParams: string = getSerializedCustomParamsResult.value;

        const createLambdaResult: Result<KurtosisLambda, Error> = this.configurator.parseParamsAndCreateKurtosisLambda(serializedCustomParams);
        if (createLambdaResult.isErr()) {
            return err(createLambdaResult.error);
        }

        const getApiContainerSocketResult: Result<string, Error> = KurtosisLambdaExecutor.getEnvVar(API_CONTAINER_SOCKET_ENV_VAR, "the socket value used in API container connection");
        if (getApiContainerSocketResult.isErr()) {
            return err(getApiContainerSocketResult.error);
        }
        const apiContainerSocket = getApiContainerSocketResult.value;

        // TODO Wrap in exception-handling???
        const apiClient: ApiContainerServiceClient = new ApiContainerServiceClient(apiContainerSocket, grpc.credentials.createInsecure());
        const networkCtx: NetworkContext = new NetworkContext(
            apiClient,
            EXECUTION_VOLUME_MOUNTPOINT
        );

        // TODO
        const lambdaServiceServer: ILambdaServiceServer;
        // lambdaServiceServer := NewKurtosisLambdaServiceServer(lambda, networkCtx)
        const serviceRegistrationFuncs: { (server: TypedServerOverride): void; }[] = [
            (server: TypedServerOverride) => {
                server.addTypedService(LambdaServiceService, lambdaServiceServer);
            }
        ]

        const lambdaServer: MinimalGRPCServer = new MinimalGRPCServer(
            LISTEN_PORT,
            GRPC_SERVER_STOP_GRACE_PERIOD_SECONDS,
            serviceRegistrationFuncs
        );
        const runResult: Result<null, Error> = await lambdaServer.run();
        if (runResult.isErr()) {
            return err(runResult.error);
        }

        return ok(null);
    }

    // ====================================================================================================
    //                                       Private helper functions
    // ====================================================================================================
    private static getEnvVar(envVarName: string, envVarDescription: string): Result<string, Error> {
        if (!(SERIALIZED_CUSTOM_PARAMS_ENV_VAR in process.env)) {
            return err(new Error("Expected an '" + envVarName + "' environment variable containing '" + envVarDescription + "', but none was found"));

        }
        const envVarValue = process.env[envVarName];
        if (envVarValue === "") {
            return err(new Error("The '" + envVarName + "' environment variable was defined, but is emptystring"));
        }

        return ok(envVarValue);
    }
}